### DISCUSSION TITLE
How do you gracefully exit in a collect_modifyitems hook ?

### DISCUSSION THREAD

**Discussion by symonk** [pytest-dev/pytest#8077] [Q&A] [answered]
I have a scenario where I need the collected items, as I want to dynamically apply some behaviour based on markers, however if something goes wrong inside `pytest_collection_modifyitems` what is the best way to gracefully terminate the run?  Especially when xdist is involved.  I have something like this at the moment, is it the best way?

Pseudo Hook:

```python
    @pytest.hookimpl
    def pytest_collection_modifyitems(self, session: Session, items: List[Item]) -> None:
        try:
            do_something()
        except SomeException as exc:
            session.shouldstop = f"something bad happened: {str(exc)}"
            pytest.exit(str(exc))
```

Is this sufficient? I find when i just pytest.exit() here, im getting a 1 exit code (without a `session.shouldstop`), I want/need to exit with a 2/3 here, is setting shouldstop explicitly like this acceptable?  This plugin is only registered on master in an xdist setup, i assume the exit() call will close down all workers too.

**Comment by symonk** [comment, marked as answer]
actually ignore this, `pytest.exit()` in the hook returns a 2 not a 1 :)

**Reply by RonnyPfannschmidt** [reply]
marking this as the answer then, thanks for the followup
